crates/sui-framework/sources/collectible.move

```diff
@@ -0,0 +1,446 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Module presenting an implementable `Collectible` type.
+/// It serves as a type-unifier and a generic interface for collectibles
+/// by allowing type-param-specific implemntations for the common type.
+///
+/// Reference implementation can be found in tests.
+///
+/// TODO:
+///  - should there be a `burn` method? does it decrease max_supply?
+///  - example implementation in Sui Framework or Examples (DevNetNFT?)
+module sui::collectible {
+    use sui::tx_context::{TxContext};
+    use std::option::{Self, Option};
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+
+    /// For when owner tries to mint more than max_supply.
+    const EMaxSupplyReached: u64 = 1;
+
+    /// Generic type for collectibles.
+    /// Combination of a `store` and `drop` capabilities saves
+    /// from locking assets inside `Collectible` while making it possible
+    /// to use `T` as both a witness and a data type.
+    struct Collectible<T: store + drop> has key {
+        id: UID,
+        /// Incremental number in a collection.
+        unique_number: u64,
+        /// Total number of items in collection. Improves
+        /// on-chain discovery (eg when Cap is wrapped).
+        edition: Option<u64>,
+        /// Optional field for custom metadata.
+        /// The same T is used for witness-ing transfer* methods.
+        info: T,
+    }
+
+    /// A capability that allows the owner to mint new `Collectible` objects. Acts as
+    /// a `TreasuryCap` for `Collectible`s.
+    struct CollectionManagerCap<phantom T: store + drop> has key, store {
+        id: UID,
+        total_supply: u64,
+        max_supply: Option<u64>
+    }
+
+
+    /// Borrow `unique_number` field.
+    public fun unique_number<T: store + drop>(self: &Collectible<T>): u64 { self.unique_number }
+
+    /// Borrow `data` field.
+    public fun info<T: store + drop>(self: &Collectible<T>): &T { &self.info }
+
+    /// Implementable method giving a mutable reference to `data`.
+    public fun info_mut<T: store + drop>(_w: T, self: &mut Collectible<T>): &mut T { &mut self.info }
+
+    /// Get ID for pointing to the `Collectible`.
+    public fun id<T: store + drop>(self: &Collectible<T>): ID { object::uid_to_inner(&self.id) }
+
+    /// Get a mutable reference to UID for allowing `transfer_to_object_id` with `Collectible` type.
+    ///
+    /// `Collectible` does not have `store` ability and it means that `transfer_to_object` call can
```

Comment:
this uses 'store' ability but above it is just key ability, nice to be consistent